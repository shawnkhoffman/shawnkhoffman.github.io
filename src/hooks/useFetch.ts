import { useState, useEffect } from 'react';

export const useFetch = <T>(fetchFunction: () => Promise<T>) => {
  const [data, setData] = useState<T | null>(null);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const result = await fetchFunction();
        setData(result);
      } catch (err) {
        console.error('Fetch error:', err);
        setError('Error loading data');
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [fetchFunction]);

  return { data, loading, error };
};