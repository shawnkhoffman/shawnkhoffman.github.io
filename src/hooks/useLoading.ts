import { useState, useEffect } from 'react';

export const useLoading = (duration: number = 500) => {
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const timer = setTimeout(() => setIsLoading(false), duration);
    return () => clearTimeout(timer);
  }, [duration]);

  return isLoading;
};