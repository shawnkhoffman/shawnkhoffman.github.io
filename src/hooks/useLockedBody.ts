import { useEffect, useCallback, useRef } from 'react';

interface OriginalStyle {
  overflow: string;
  paddingRight: string;
  position: string;
  top: string;
  left: string;
  right: string;
  bottom: string;
}

const useLockedBody = (locked: boolean = true): void => {
  const originalStyleRef = useRef<OriginalStyle>({
    overflow: '',
    paddingRight: '',
    position: '',
    top: '',
    left: '',
    right: '',
    bottom: '',
  });

  const getScrollbarWidth = useCallback(() => {
    const scrollDiv = document.createElement('div');
    scrollDiv.style.cssText = `
      width: 99px;
      height: 99px;
      overflow: scroll;
      position: absolute;
      top: -9999px;
    `;
    document.body.appendChild(scrollDiv);
    const scrollbarWidth = scrollDiv.offsetWidth - scrollDiv.clientWidth;
    document.body.removeChild(scrollDiv);
    return scrollbarWidth;
  }, []);

  useEffect(() => {
    const bodyStyle = document.body.style;
    originalStyleRef.current = {
      overflow: bodyStyle.overflow,
      paddingRight: bodyStyle.paddingRight,
      position: bodyStyle.position,
      top: bodyStyle.top,
      left: bodyStyle.left,
      right: bodyStyle.right,
      bottom: bodyStyle.bottom,
    };
  }, []);

  useEffect(() => {
    const bodyStyle = document.body.style;

    if (locked) {
      const scrollbarWidth = getScrollbarWidth();
      const hasVerticalScroll = window.innerWidth > document.documentElement.clientWidth;

      const scrollY = window.scrollY;

      Object.assign(bodyStyle, {
        overflow: 'hidden',
        position: 'fixed',
        top: `-${scrollY}px`,
        left: '0',
        right: '0',
        bottom: '0',
        paddingRight: hasVerticalScroll ? `${scrollbarWidth}px` : '0',
      });

      return () => {
        Object.assign(bodyStyle, originalStyleRef.current);
        window.scrollTo(0, scrollY);
      };
    } else {
      Object.assign(bodyStyle, originalStyleRef.current);
      return undefined;
    }
  }, [locked, getScrollbarWidth]);
};

export default useLockedBody;