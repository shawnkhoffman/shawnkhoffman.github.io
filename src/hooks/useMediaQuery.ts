import { useState, useEffect } from 'react';

const getMatches = (query: string): boolean => {
  if (typeof window !== 'undefined' && window.matchMedia) {
    return window.matchMedia(query).matches;
  }
  return false;
};

export default function useMediaQuery(query: string): boolean {
  const [matches, setMatches] = useState<boolean>(getMatches(query));

  useEffect(() => {
    if (typeof window !== 'undefined' && window.matchMedia) {
      const media = window.matchMedia(query);
      const listener = () => setMatches(media.matches);
      
      media.addEventListener('change', listener);
      return () => media.removeEventListener('change', listener);
    }
  }, [query]);

  return matches;
}
