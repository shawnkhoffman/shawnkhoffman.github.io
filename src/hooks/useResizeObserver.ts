import { useEffect, RefObject } from 'react';

type ObserverCallback = () => void;

const useResizeObserver = (
  ref: RefObject<Element | HTMLElement | null>,
  callback: ObserverCallback,
  options: ResizeObserverOptions = {}
) => {
  useEffect(() => {
    if (!ref.current) return;

    const element = ref.current;
    const observer = new ResizeObserver(() => {
      callback();
    });

    observer.observe(element, options);

    return () => {
      observer.disconnect();
    };
  }, [ref, callback, options]);
};

export default useResizeObserver;