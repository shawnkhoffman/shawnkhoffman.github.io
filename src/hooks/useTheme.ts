import { createContext, useContext, Dispatch, SetStateAction } from 'react';

export type Theme = 'light' | 'dark' | 'system';

interface ThemeContextProps {
  theme: Theme;
  setTheme: Dispatch<SetStateAction<Theme>>;
}

export const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) throw new Error('useTheme must be used within a ThemeProvider');
  return context;
};