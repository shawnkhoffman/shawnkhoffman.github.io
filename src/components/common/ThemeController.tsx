import { memo, useCallback, useEffect, useMemo, useState } from "react";
import { useTheme } from "../../hooks/useTheme";
import {
  FaSun,
  FaMoon,
  FaDesktop,
  FaMobileAlt,
  FaTabletAlt,
} from "react-icons/fa";
import { Theme } from "@/types/theme";

const useMediaQuery = (query: string): boolean => {
  const getMatches = (mediaQuery: string): boolean => {
    if (typeof window !== "undefined") {
      return window.matchMedia(mediaQuery).matches;
    }
    return false;
  };

  const [matches, setMatches] = useState(getMatches(query));

  useEffect(() => {
    if (typeof window === "undefined") {
      return;
    }

    const mediaQuery = window.matchMedia(query);

    const handleChange = (e: MediaQueryListEvent) => {
      setMatches(e.matches);
    };

    mediaQuery.addEventListener("change", handleChange);

    return () => {
      mediaQuery.removeEventListener("change", handleChange);
    };
  }, [query]);

  return matches;
};

export type DeviceType = "desktop" | "mobile" | "tablet";

export interface ThemeControllerProps {
  showLabel?: boolean;
  className?: string;
  labelPosition?: "left" | "right";
  size?: "sm" | "md" | "lg";
  showTooltip?: boolean;
  animated?: boolean;
}

const THEME_ORDER: Theme[] = ["light", "dark", "system"];

const DEVICE_BREAKPOINTS = {
  mobile: 768,
  tablet: 1024,
} as const;

const ICON_SIZES = {
  sm: "w-4 h-4",
  md: "w-5 h-5",
  lg: "w-6 h-6",
} as const;

const BUTTON_SIZES = {
  sm: "p-1.5",
  md: "p-2",
  lg: "p-2.5",
} as const;

const ANIMATION_CLASSES = {
  light: "animate-rotate-sun",
  dark: "animate-rotate-moon",
  system: "animate-rotate-system",
} as const;

const THEME_LABELS = {
  light: "Light theme",
  dark: "Dark theme",
  system: "System theme",
} as const;

const DEVICE_ICONS = {
  mobile: FaMobileAlt,
  tablet: FaTabletAlt,
  desktop: FaDesktop,
} as const;

const ThemeController = memo<ThemeControllerProps>(
  ({
    showLabel = false,
    className = "",
    labelPosition = "right",
    size = "md",
    showTooltip = true,
    animated = true,
  }) => {
    const { theme, setTheme } = useTheme();
    const prefersReducedMotion = useMediaQuery(
      "(prefers-reduced-motion: reduce)"
    );
    const prefersDark = useMediaQuery("(prefers-color-scheme: dark)");
    const isMobile = useMediaQuery(
      `(max-width: ${DEVICE_BREAKPOINTS.mobile}px)`
    );
    const isTablet = useMediaQuery(
      `(max-width: ${DEVICE_BREAKPOINTS.tablet}px)`
    );

    const deviceType = useMemo((): DeviceType => {
      if (isMobile) return "mobile";
      if (isTablet) return "tablet";
      return "desktop";
    }, [isMobile, isTablet]);

    const iconSize = useMemo(() => ICON_SIZES[size], [size]);
    const buttonSize = useMemo(() => BUTTON_SIZES[size], [size]);

    const shouldAnimate = useMemo(
      () => animated && !prefersReducedMotion,
      [animated, prefersReducedMotion]
    );

    const getNextTheme = useCallback((currentTheme: Theme): Theme => {
      const currentIndex = THEME_ORDER.indexOf(currentTheme);
      const nextIndex = (currentIndex + 1) % THEME_ORDER.length;
      return THEME_ORDER[nextIndex];
    }, []);

    const handleThemeChange = useCallback(() => {
      setTheme(getNextTheme(theme as Theme));
    }, [theme, setTheme, getNextTheme]);

    useEffect(() => {
      if (theme === "system") {
        document.documentElement.setAttribute(
          "data-theme",
          prefersDark ? "dark" : "light"
        );
      }
    }, [theme, prefersDark]);

    const renderIcon = useCallback(() => {
      const animationClass = shouldAnimate
        ? ANIMATION_CLASSES[theme as Theme] || ""
        : "";
      const baseClasses = `${iconSize} transition-transform duration-200`;
      const combinedClasses = `${baseClasses} ${animationClass}`;

      switch (theme) {
        case "light":
          return (
            <FaSun
              className={`text-content ${combinedClasses}`}
              aria-hidden="true"
            />
          );
        case "dark":
          return (
            <FaMoon
              className={`text-content ${combinedClasses}`}
              aria-hidden="true"
            />
          );
        case "system": {
          const DeviceIcon = DEVICE_ICONS[deviceType];
          return (
            <DeviceIcon
              className={`text-info ${combinedClasses}`}
              aria-hidden="true"
            />
          );
        }
        default:
          return null;
      }
    }, [theme, deviceType, shouldAnimate, iconSize]);

    const buttonContent = useMemo(
      () => (
        <>
          {renderIcon()}
          {showLabel && (
            <span
              className={`text-sm font-medium ${
                labelPosition === "left" ? "ml-2" : "mr-2"
              }`}
            >
              Theme
            </span>
          )}
        </>
      ),
      [showLabel, labelPosition, renderIcon]
    );

    const handleKeyDown = useCallback(
      (event: React.KeyboardEvent) => {
        if (event.key === "Enter" || event.key === " ") {
          event.preventDefault();
          handleThemeChange();
        }
      },
      [handleThemeChange]
    );

    return (
      <button
        className={`
        inline-flex items-center justify-center rounded-full
        hover:bg-base-200 focus:outline-none focus:ring-2
        focus:ring-offset-2 focus:ring-info
        ${buttonSize}
        ${className}
      `}
        onClick={handleThemeChange}
        onKeyDown={handleKeyDown}
        aria-label={`Switch theme to ${getNextTheme(theme as Theme)} mode`}
        title={showTooltip ? THEME_LABELS[theme as Theme] : undefined}
        data-theme-controller
        role="switch"
        aria-checked={theme === "dark"}
        data-testid={`theme-switch-${theme}`}
      >
        <div className="flex items-center justify-center relative">
          {buttonContent}
        </div>
      </button>
    );
  }
);

ThemeController.displayName = "ThemeController";

export default ThemeController;
