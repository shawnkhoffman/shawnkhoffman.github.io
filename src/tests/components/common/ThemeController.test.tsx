import { describe, it, expect, vi, afterEach } from 'vitest';
import ThemeController from '../../../components/common/ThemeController';
import '@testing-library/jest-dom';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import { ThemeContext } from '../../../hooks/useTheme';
import type { Theme } from '../../../types/theme';

const localStorageMock = {
  getItem: vi.fn(),
  setItem: vi.fn(),
  clear: vi.fn()
};
Object.defineProperty(window, 'localStorage', { value: localStorageMock });

describe('ThemeController', () => {
  const setup = (initialTheme: Theme = 'light', props = {}) => {
    cleanup();
    
    const mockSetTheme = vi.fn();
    return {
      mockSetTheme,
      ...render(
        <ThemeContext.Provider value={{ theme: initialTheme, setTheme: mockSetTheme }}>
          <ThemeController {...props} />
        </ThemeContext.Provider>
      ),
      getButton: () => screen.getByRole('switch')
    };
  };

  afterEach(cleanup);

  describe('Rendering', () => {
    it('renders consistently across viewport sizes', () => {
      window.matchMedia = vi.fn().mockImplementation(query => ({
        matches: query.includes('768'),
        media: query,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      }));
      const { container: mobile } = setup();
      expect(mobile).toMatchSnapshot('mobile');

      window.matchMedia = vi.fn().mockImplementation(() => ({
        matches: false,
        media: '',
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      }));
      const { container: desktop } = setup();
      expect(desktop).toMatchSnapshot('desktop');
    });
  });

  describe('Structure', () => {
    it('contains required elements', () => {
      const { getButton } = setup();
      const button = getButton();
      expect(button).toBeInTheDocument();
      expect(button.querySelector('[aria-hidden="true"]')).toBeInTheDocument();
    });
  });

  describe('Interactions', () => {
    it('handles theme cycling correctly', () => {
      const { mockSetTheme, getButton } = setup('light');
      const button = getButton();

      fireEvent.click(button);
      expect(mockSetTheme).toHaveBeenCalledWith('dark');

      fireEvent.keyDown(button, { key: 'Enter' });
      fireEvent.keyDown(button, { key: ' ' });
      expect(mockSetTheme).toHaveBeenCalledTimes(3);
    });
  });

  describe('Accessibility', () => {
    it('meets accessibility requirements', () => {
      const { getButton } = setup();
      const button = getButton();
      
      expect(button).toHaveAttribute('role', 'switch');
      expect(button).toHaveAttribute('aria-label');
      expect(button).toHaveAttribute('aria-checked');
    });

    it('respects user preferences', () => {
      window.matchMedia = vi.fn().mockImplementation(query => ({
        matches: query.includes('reduced-motion'),
        media: query,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      }));
      const { container } = setup();
      expect(container).toMatchSnapshot('reduced-motion');
    });
  });
});
