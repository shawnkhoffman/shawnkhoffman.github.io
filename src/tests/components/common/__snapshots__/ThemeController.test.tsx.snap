// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeController > Accessibility > respects user preferences > desktop 1`] = `
<div>
  <button
    aria-checked="false"
    aria-label="Switch theme to dark mode"
    class="
        inline-flex items-center justify-center rounded-full
        hover:bg-base-200 focus:outline-none focus:ring-2
        focus:ring-offset-2 focus:ring-info
        p-2
        
      "
    data-testid="theme-switch-light"
    data-theme-controller="true"
    role="switch"
    title="Light theme"
  >
    <div
      class="flex items-center justify-center relative"
    >
      <svg
        aria-hidden="true"
        class="text-content w-5 h-5 transition-transform duration-200 animate-rotate-sun"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
        />
      </svg>
    </div>
  </button>
</div>
`;

exports[`ThemeController > Accessibility > respects user preferences > mobile 1`] = `
<div>
  <button
    aria-checked="false"
    aria-label="Switch theme to dark mode"
    class="
        inline-flex items-center justify-center rounded-full
        hover:bg-base-200 focus:outline-none focus:ring-2
        focus:ring-offset-2 focus:ring-info
        p-2
        
      "
    data-testid="theme-switch-light"
    data-theme-controller="true"
    role="switch"
    title="Light theme"
  >
    <div
      class="flex items-center justify-center relative"
    >
      <svg
        aria-hidden="true"
        class="text-content w-5 h-5 transition-transform duration-200 animate-rotate-sun"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
        />
      </svg>
    </div>
  </button>
</div>
`;

exports[`ThemeController > Accessibility > respects user preferences > reduced-motion 1`] = `
<div>
  <button
    aria-checked="false"
    aria-label="Switch theme to dark mode"
    class="
        inline-flex items-center justify-center rounded-full
        hover:bg-base-200 focus:outline-none focus:ring-2
        focus:ring-offset-2 focus:ring-info
        p-2
        
      "
    data-testid="theme-switch-light"
    data-theme-controller="true"
    role="switch"
    title="Light theme"
  >
    <div
      class="flex items-center justify-center relative"
    >
      <svg
        aria-hidden="true"
        class="text-content w-5 h-5 transition-transform duration-200 "
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
        />
      </svg>
    </div>
  </button>
</div>
`;
